---
title: "IFNb_ScRNAseq_Integrated_Analysis"
output: rmarkdown::github_document
---

## Objectives

### To assess cell-type-specific changes in gene expression with IFN-beta treatment

### Peripheral blood mononuclear cells (PBMCs) from eight lupus patients were split into a stimulated and control group and the stimulated group was treated with interferone beta.

### To integrate data by conditions; to overlay cells that are similar in both conditions

## Data

### Raw data is accessible from 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Setup

```{r}
pacman::p_load(tidyverse, harmony, Seurat, SeuratData, ggplot2, patchwork)



rm(list=ls())
```

## Read in data and Create Seurat Object

```{r, results='hide'}
# install dataset
InstallData("ifnb")

# load dataset
LoadData("ifnb")
str(ifnb)

```

## Calculate % mitochondria (dying cells)

```{r, results='hide'}
ifnb.updated = UpdateSeuratObject(object = ifnb)
ifnb.updated$mito.percent <- PercentageFeatureSet(ifnb.updated, pattern = '^MT-')
View(ifnb.updated@meta.data)
```

## Filtering

```{r}
ifnb.updated
ifnb.filtered <- subset(ifnb.updated, subset = nCount_RNA > 800 &
                          nFeature_RNA > 200 & 
                          mito.percent < 5)

```

## Data Preprocessing and Visualize data before integration to correct for batch effects

```{r, warning=FALSE}
# perform standard workflow steps 
ifnb.filtered <- NormalizeData(ifnb.filtered)
ifnb.filtered <- FindVariableFeatures(ifnb.filtered)
ifnb.filtered <- ScaleData(ifnb.filtered)
ifnb.filtered <- RunPCA(ifnb.filtered)
ElbowPlot(ifnb.filtered)
ifnb.filtered <- RunUMAP(ifnb.filtered, dims = 1:20, reduction = 'pca')

before <- DimPlot(ifnb.filtered, reduction = 'umap', group.by = 'stim')
before
```

## Run Harmony and Get Harmony Embeddings

```{r, results='hide'}
ifnb.harmony <- ifnb.filtered %>%
  RunHarmony(group.by.vars = 'stim', plot_convergence = FALSE)

ifnb.harmony@reductions

ifnb.harmony.embed <- Embeddings(ifnb.harmony, "harmony")
ifnb.harmony.embed[1:10,1:10]
```

## Scale data, Run UMAP to integrate the data

```{r, results='hide'}
ifnb.harmony <- ifnb.harmony %>%
  RunUMAP(reduction = 'harmony', dims = 1:20) %>%
  FindNeighbors(reduction = "harmony", dims = 1:20) %>%
  FindClusters(resolution = 0.5)
```

## Visualize integrated data

```{r}
# visualize 
after <- DimPlot(ifnb.harmony, reduction = 'umap', group.by = 'stim')

before|after
```

## Save harmony Object

```{r}
saveRDS(ifnb.harmony, file = "./results/ifnb_harmony.rds")
```

## Visualize the data

```{r}
# visualize data
clusters <- DimPlot(ifnb.harmony, reduction = 'umap', group.by = 'seurat_clusters', label = TRUE)
condition <- DimPlot(ifnb.harmony, reduction = 'umap', group.by = 'stim')
condition|clusters
```

## FindConserved markers

```{r, results='hide'}
# Notes:
# slot depends on the type of the test used, 
# default is data slot that stores normalized data
# DefaultAssay(ifnb_harmony) <- 'RNA'

DefaultAssay(ifnb.harmony) <- 'RNA'

markers_cluster3 <- FindConservedMarkers(ifnb.harmony,
                     ident.1 = 3,
                     grouping.var = 'stim')

head(markers_cluster3)
```

## Visualize the top features from Cluster 3 of conserved markers

```{r}
# let's visualize top features
FeaturePlot(ifnb.harmony, features = c('FCGR3A', 'VMO1', 'MS4A7', 'CXCL16'), min.cutoff = 'q10')


# min-cut off explanation:
seq(1,5)
SetQuantile('q50', seq(1,5))
SetQuantile('q10', seq(1,5))
```

## Annotate the different clusters

```{r, results='hide'}
# cells already have annotations provided in the metadata
View(ifnb.harmony@meta.data)

# Settings cluster identities is an iterative step
# multiple approaches could be taken - automatic/manual anotations (sometimes both)
# need to make sure each cell type forms a separate cluster

# setting Idents as Seurat annotations provided (also a sanity check!)
Idents(ifnb.harmony) <- ifnb.harmony@meta.data$seurat_annotations
Idents(ifnb.harmony)
```

```{r}
DimPlot(ifnb.harmony, reduction = 'umap', label = TRUE)
```

## FindMarkers between conditions

```{r, results='hide'}
ifnb.harmony$celltype.stim <- paste0(ifnb.harmony$seurat_annotations,'_', ifnb.harmony$stim)
ifnb.harmony$celltype <- Idents(ifnb.harmony)
Idents(ifnb.harmony) <- "celltype.stim"
View(ifnb.harmony@meta.data)
```


```{r}
DimPlot(ifnb.harmony, reduction = 'umap', label = TRUE)
```

## Find markers

```{r, results='hide'}
b.interferon.response <- FindMarkers(ifnb.harmony, ident.1 = 'CD16 Mono_STIM', ident.2 = 'CD16 Mono_CTRL')

head(b.interferon.response)
```

## Differentially expressed genes (by conditions)

```{r}
FeaturePlot(ifnb.harmony, features = c('IFIT1', 'IFI6', 'ISG15'), split.by = 'stim', min.cutoff = 'q10')
```

## plotting conserved features vs DE features between conditions (by conditions)

```{r, results='hide'}
head(markers_cluster3)

FeaturePlot(ifnb.harmony, features = c('FCGR3A', 'ISG15'), split.by = 'stim', min.cutoff = 'q10')
```


## plotting conserved features vs DE features between conditions (by conditions and cell types)

```{r}

plots <- VlnPlot(ifnb.harmony, features = c('MS4A7', 'CXCL16', 'IFI6', 'IFIT1'), split.by = "stim", group.by = "celltype", 
    pt.size = 0, combine = FALSE)
wrap_plots(plots = plots, ncol = 1)
```

